src/main/java/org/spongepowered/common/effect/particle/SpongeParticleHelper.java

```diff
@@ -75,6 +85,40 @@ private static int getBlockState(SpongeParticleEffect effect, Optional<BlockStat
         }
     }
 
+    private static int getDirectionData(Direction direction) {
+        if (direction.isSecondaryOrdinal()) {
+            direction = Direction.getClosest(direction.asOffset(), Direction.Division.CARDINAL);
+        }
+        switch (direction) {
+            case SOUTHEAST:
+                return 0;
+            case SOUTH:
+                return 1;
+            case SOUTHWEST:
+                return 2;
+            case EAST:
+                return 3;
+            case WEST:
+                return 5;
+            case NORTHEAST:
+                return 6;
+            case NORTH:
+                return 7;
+            case NORTHWEST:
+                return 8;
+            default:
+                return 4;
+        }
+    }
+
+    private static final EntityFireworkRocket FIREWORK_ROCKET_DUMMY = new EntityFireworkRocket(null);
```

Comment:
I understand that, but why are you caching the packets and modifying the data, and then sending? I can only see problems doing this - being modified again before being sent with different data, etc.